host/crypto/bcrypt/cert.c

```diff
@@ -0,0 +1,1295 @@
+    *urls_count = found_urls_count;
+    *urls = found_urls;
+    found_urls = NULL;
+    result = OE_OK;
+
+done:
+    _free_urls_array(&found_urls, &found_urls_count);
+
+    return result;
+}
+
+/*
+**==============================================================================
+**
+** Public functions
+**
+**==============================================================================
+*/
+
+/* Used by tests/crypto_crls_cert_chains */
+oe_result_t oe_cert_read_pem(
+    oe_cert_t* cert,
+    const void* pem_data,
+    size_t pem_data_size)
+{
+    oe_result_t result = OE_UNEXPECTED;
+    cert_t* impl = (cert_t*)cert;
+    BYTE* der_data = NULL;
+    DWORD der_data_size = 0;
+
+    /* Zero-initialize the implementation */
+    if (impl)
+        memset(impl, 0, sizeof(cert_t));
+
+    OE_CHECK(oe_bcrypt_pem_to_der(
+        pem_data, pem_data_size, &der_data, &der_data_size));
+    OE_CHECK(oe_cert_read_der(cert, der_data, der_data_size));
+
+    result = OE_OK;
+
+done:
+    if (der_data)
+        free(der_data);
+
+    return result;
+}
+
+oe_result_t oe_cert_read_der(
+    oe_cert_t* cert,
+    const void* der_data,
+    size_t der_data_size)
+{
+    oe_result_t result = OE_UNEXPECTED;
+    cert_t* impl = (cert_t*)cert;
+    PCCERT_CONTEXT cert_context = NULL;
+
+    /* Zero-initialize the implementation */
+    if (impl)
+        memset(impl, 0, sizeof(cert_t));
+
+    if (der_data_size > MAXDWORD)
+        OE_RAISE(OE_INVALID_PARAMETER);
+
+    /* Create the CERT_CONTEXT from DER data */
+    cert_context = CertCreateCertificateContext(
+        X509_ASN_ENCODING | PKCS_7_ASN_ENCODING,
+        der_data,
+        (DWORD)der_data_size);
+
+    if (!cert_context)
+        OE_RAISE_MSG(
+            OE_CRYPTO_ERROR,
+            "CertCreateCertificateContext failed, err=%#x\n",
+            GetLastError());
+
+    /* Initialize the wrapper cert_context structure */
+    _cert_init(impl, cert_context);
+    cert_context = NULL;
+    result = OE_OK;
+
+done:
+    if (cert_context)
+        CertFreeCertificateContext(cert_context);
+
+    return result;
+}
+
+oe_result_t oe_cert_free(oe_cert_t* cert)
+{
+    oe_result_t result = OE_UNEXPECTED;
+    cert_t* impl = (cert_t*)cert;
+
+    /* Check parameters */
+    if (!_cert_is_valid(impl))
+        OE_RAISE(OE_INVALID_PARAMETER);
+
+    /* Free the certificate */
+    CertFreeCertificateContext(impl->cert);
+    _cert_destroy(impl);
+
+    result = OE_OK;
+
+done:
+    return result;
+}
+
+oe_result_t oe_cert_chain_read_pem(
+    oe_cert_chain_t* chain,
+    const void* pem_data,
+    size_t pem_data_size)
+{
+    oe_result_t result = OE_UNEXPECTED;
+    cert_chain_t* impl = (cert_chain_t*)chain;
+    PCCERT_CHAIN_CONTEXT cert_chain = NULL;
+    HCERTSTORE cert_store = NULL;
+    uint32_t cert_count = 0;
+    PCCERT_CONTEXT cert_context = NULL;
+
+    /* Zero-initialize the implementation */
+    if (impl)
+        memset(impl, 0, sizeof(cert_chain_t));
```

Comment:
Would memset_s be appropriate here?

Reply:
Turns out `memset_s` is actually not one of MSVC's [Security-Enhanced Versions of CRT Functions](https://docs.microsoft.com/en-us/cpp/c-runtime-library/security-enhanced-versions-of-crt-functions?view=vs-2019) unlike `memcpy_s`